rust/src/mqtt/logger.rs

```diff
@@ -0,0 +1,250 @@
+                js.set("subscribe", msg_json);
+            }
+            MQTTOperation::SUBACK(ref suback) => {
+                msg_json.set_integer("message_id", suback.message_id as u64);
+                let qos_json = Json::array();
+                for t in &suback.qoss {
+                    qos_json.array_append_integer(*t as u64);
+                }
+                msg_json.set("qos_granted", qos_json);
+                js.set("suback", msg_json);
+            }
+            MQTTOperation::UNSUBSCRIBE(ref unsub) => {
+                msg_json.set_integer("message_id", unsub.message_id as u64);
+                let unsub_json = Json::array();
+                for t in &unsub.topics {
+                    unsub_json.array_append_string(t);
+                }
+                msg_json.set("topics", unsub_json);
+                js.set("unsubscribe", msg_json);
+            }
+            MQTTOperation::UNSUBACK(ref unsuback) => {
+                msg_json.set_integer("message_id", unsuback.message_id as u64);
+                if let Some(codes) = &unsuback.reason_codes {
+                    let rcodes_json = Json::array();
+                    for t in codes {
+                        rcodes_json.array_append_integer(*t as u64);
+                    }
+                    msg_json.set("reason_codes", rcodes_json);
+                }
+                js.set("unsuback", msg_json);
+            }
+            MQTTOperation::AUTH(ref auth) => {
+                msg_json.set_integer("reason_code", auth.reason_code as u64);
+                if let Some(properties) = &auth.properties {
+                    let prop_json = Json::object();
+                    for prop in properties {
+                        prop.set_json(&prop_json);
+                    }
+                    msg_json.set("properties", prop_json);
+                }
+                js.set("auth", msg_json);
+            }
+            MQTTOperation::DISCONNECT(ref disco) => {
+                if let Some(reason_code) = &disco.reason_code {
+                    msg_json.set_integer("reason_code", *reason_code as u64);
+                }
+                if let Some(properties) = &disco.properties {
+                    let prop_json = Json::object();
+                    for prop in properties {
+                        prop.set_json(&prop_json);
+                    }
+                    msg_json.set("properties", prop_json);
+                }
+                js.set("disconnect", msg_json)
+            },
+            ref _what => {
+                js.set("unassigned", msg_json);
+            }
+        }
+    }
+    return Some(js);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_logger_log(_state: &mut MQTTState, tx: *mut std::os::raw::c_void, flags: u32) -> *mut JsonT {
+    let tx = cast_pointer!(tx, MQTTTransaction);
+    match log_mqtt(tx, flags) {
+        Some(js) => js.unwrap(),
+        None => std::ptr::null_mut(),
+    }
```

Comment:
None is never returned, is it ?

Reply:
This is actually obsolete after switch to JsonBuilder.